### PR TITLE
Browser sample

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

*Description of changes:*
Added browser pub-sub sample. The update depends on aws-crt changes on https://github.com/awslabs/aws-crt-nodejs/pull/319
**TODO: update crt dependency before merge**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samples/README.md [hunk 1]
```diff
@@ -99,6 +99,18 @@ npm install
 node index.js --endpoint <endpoint> --ca_file <file> --cert <file> --key <file>
 ```
 
+## Browser/pub_sub
+
+This is a browser based version of the PubSub sample in JS.  If you don't want to use browser, this
```
Comment (bretambrose) [inline samples/README.md:104]: I would drop the second sentence entirely.

File: samples/browser/pub_sub/index.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+async function connect_websocket(credentials) {
+  return new Promise((resolve, reject) => {
+    let config =
+      iot.AwsIotMqttConnectionConfigBuilder.new_builder_for_websocket()
+        .with_clean_session(true)
+        .with_client_id(`pub_sub_sample(${new Date()})`)
+        .with_endpoint(Settings.AWS_IOT_ENDPOINT)
+        .with_credentials(
```
Comment (bretambrose) [inline samples/browser/pub_sub/index.js:59]: Is it possible to update this to fetch the credentials as part of the signing handshake async flow?  Cognito credentials are session-based which means this sample will eventually go into a permanent failure mode on the first reconnect after session expiration.

Comment (bretambrose) [reply samples/browser/pub_sub/index.js:59]: We're already using aws-sdk-js in this sample.

**Comment by jfloodnet** [comment]
Hi - this is looking pretty straight forward. If the sticking point of this PR is the sample code, can you separate out the enabling changes in "browser.ts" and get that merged first? I'd like to make use of this asap. cheers.

File: samples/browser/pub_sub/index.js (hunk 2 above)
Comment (bretambrose) [reply samples/browser/pub_sub/index.js:59]: Revisiting this, I think my comment is being misinterpreted.

As it stands, this sample (conditionally) uses Cognito but the sourced credentials are "static" and so someone taking this sample directly will not have a good starting point: as soon as the connection drops the first time after credential expiration, every reconnect will fail.  It's slightly more complicated, but if the sample instead has a handshake transform callback that sources credentials and then signs with the sourced credentials, the sample will not suffer from this defect and people using it as a starting point won't run into this fundamental issue.

Comment (TingDaoK) [reply samples/browser/pub_sub/index.js:59]: I don't think the URL transform is exposed as well. I agree the current sample for Cognito is not really helpful. What about we push the sample with static credentials only first to unblock some browser customers and expose either the credential provider override or the URL transform override from crt later?

**Comment by ZeeD** [comment]
Just a question. Is there a reason the example relies on aws-sdk v2 and not on aws-sdk v3 ?

**Comment by jmklix** [comment]
@ZeeD this aws-iot-device-sdk-js-v2 is completely separate from and not dependent on the aws-sdk-js-v2 or aws-sdk-js-v3.

**Comment by ZeeD** [comment]
@jmklix  absolutely, they are two different projects.

Still, in this demo, both of them are used: https://github.com/aws/aws-iot-device-sdk-js-v2/blob/7f9c3b92309a32c03ccf15b17c96c63be0d213f2/samples/browser/pub_sub/package.json#L42-L43
(from what is my understanding aws-sdk is used for auth reasons, as the aws-iot-device doesn't offer this functionality for browsers)

And I was wondering, for the demo purpose, what was the rationale to relies on "aws-sdk": "^2.585.0"  and not on v3.

FWIW: I'm starting to work on a project where we need to connect browsers to IoT backend systems and I could use this demo as a starting point

**Comment by TingDaoK** [comment]
> @jmklix absolutely, they are two different projects.
> 
> Still, in this demo, both of them are used:
> 
> https://github.com/aws/aws-iot-device-sdk-js-v2/blob/7f9c3b92309a32c03ccf15b17c96c63be0d213f2/samples/browser/pub_sub/package.json#L42-L43
> 
> (from what is my understanding aws-sdk is used for auth reasons, as the aws-iot-device doesn't offer this functionality for browsers)
> 
> And I was wondering, for the demo purpose, what was the rationale to relies on "aws-sdk": "^2.585.0" and not on v3.
> 
> FWIW: I'm starting to work on a project where we need to connect browsers to IoT backend systems and I could use this demo as a starting point

No particular reason, you can try to use aws-sdk v3 instead. 
But, the issue for the current sample is that the credentials will NOT be updated overtime. So, if you have your browser project longs longer than the credentials valid time, it will fail.

**Comment by ZeeD** [comment]
> But, the issue for the current sample is that we only take static credentials so that the credentials will NOT be updated overtime. So, if you have your browser project longs longer than the credentials valid time, it will fail.

Ok thanks about the notice.

BTW I was toying with the demo and I saw that during the setup you can set a [credential provider](https://aws.github.io/aws-iot-device-sdk-js-v2/interfaces/aws_iot.websocketconfig.html#credentials_provider) but the api doesn't seems to support aws sdk's [credential-providers](https://www.npmjs.com/package/@aws-sdk/credential-providers).

Is this expected?
Should I write some form of adapter? If yes, apart from the sources, is there some documentation on how to write a provider and what kind of interactions should It support?

**Comment by TingDaoK** [comment]
> BTW I was toying with the demo and I saw that during the setup you can set a [credential provider](https://aws.github.io/aws-iot-device-sdk-js-v2/interfaces/aws_iot.websocketconfig.html#credentials_provider) but the api doesn't seems to support aws sdk's [credential-providers](https://www.npmjs.com/package/@aws-sdk/credential-providers).
Is this expected?
Should I write some form of adapter? If yes, apart from the sources, is there some documentation on how to write a provider and what kind of interactions should It support?

We just don't have any credential provider implemented for Browser. [That](https://awslabs.github.io/aws-crt-nodejs/browser/modules/auth.html)'s the browser doc, what you saw the credential provider doc is for node, [here](https://awslabs.github.io/aws-crt-nodejs/node/modules/auth.html). It's a feature we need to add to support any type of dynamic credentials.

**Review by sbSteveK — APPROVED** [review]

**Review by TwistedTwigleg — APPROVED** [review]
Looks good!

It will need a rebase, since when I added the custom authorizer sample I added `lib/browser.ts`. Also, I would recommend removing `package-lock.json` so we don't have extra dependencies in the samples to update but that is a super minor thing and we can do it afterwards.

Other than those two minor things, everything looks great 👍

**Pull request merged**